pkgs/top-level/all-packages.nix

```diff
@@ -16245,15 +16245,57 @@ with pkgs;
   gcc-arm-embedded-12 = callPackage ../development/compilers/gcc-arm-embedded/12 { };
   gcc-arm-embedded = gcc-arm-embedded-12;
 
-  # It would be better to match the default gcc so that there are no linking errors
-  # when using C/C++ libraries in D packages, but right now versions >= 12 are broken.
-  gdc = gdc11;
+  # Has to match the default gcc so that there are no linking errors when
+  # using C/C++ libraries in D packages
+  gdc = gdc13;
   gdc11 = wrapCC (gcc11.cc.override {
     name = "gdc";
-    langCC = false;
-    langC = false;
+    langCC = true; # required to bootstrap newer gdc
+    langC = true;
     langD = true;
     profiledCompiler = false;
+    gdc = null; # gdc>=12 requires gdc to bootstrap
+  } // {
+    # GDC doesn't support aarch64-darwin yet
+    meta.broken = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64;
+  });
+
+  gdc12 = wrapCC (gcc12.cc.override {
+    name = "gdc";
+    langCC = true;
+    langC = true;
+    langD = true;
+    profiledCompiler = false;
+    gdc = gdc11; # gdc>=12 requires gdc to bootstrap
+    stdenv =
+      if stdenv.hostPlatform == stdenv.targetPlatform
+         && stdenv.buildPlatform == stdenv.hostPlatform
+         && stdenv.buildPlatform.isDarwin
+         && stdenv.buildPlatform.isx86_64
+      then overrideCC stdenv gdc11
+      else stdenv;
+  } // {
+    # GDC doesn't support aarch64-darwin yet
+    meta.broken = stdenv.hostPlatform.isDarwin && stdenv.hostPlatform.isAarch64;
+  });
+
+  gdc13 = wrapCC (gcc13.cc.override {
+    name = "gdc";
+    langCC = true;
+    langC = true;
+    langD = true;
+    profiledCompiler = false;
+    gdc = gdc11; # gdc>=12 requires gdc to bootstrap
+    stdenv =
```

Comment:
This was copied from gnat's x86_64-darwin support (#209423) but I don't know why it doesn't work without. @boeyms would you be able to help?

Reply:
@emilytrau I can tell you why `overrideCC stdenv gnatboot12` was used to build `gnat12` on x86_84-darwin; I'd have to look into the gdc bootstrap process to see if the same reasons apply here.

The key part is what is explained in https://github.com/NixOS/nixpkgs/pull/209423/commits/82b88d2db629a06443a6d7befb8518dbb785d183 from https://github.com/NixOS/nixpkgs/pull/209423, both in the commit message and at https://github.com/NixOS/nixpkgs/pull/209423/commits/82b88d2db629a06443a6d7befb8518dbb785d183?diff=unified&w=0#diff-5bb3264ab3e5dc044855e48c8c2c1ac4b7883da5de9db2a4c4e094030dee7c2cR62-R80.  However, it may help if I explain things differently and more fully here.

- For NixOS on x86_64-darwin, at least at the time I created that PR, we had two assemblers to worry about.
  - The assembler from the Nixpkgs `cctools-port` package, which is used only on Darwin, and which Nixpkgs used as part of its default Darwin stdenv.  This turns out to be a fork of a very old version (1.x) of the GNU assembler, and was originally forked by Apple to serve as the assembler for the original versions of Mac OS X and Xcode.
  - The LLVM/Clang integrated assembler that now comes with Xcode, which replaced the above assembler in Xcode once Apple considered it mature enough.
- It so happens that the two assemblers use incompatible syntax for some x86-64 instructions, such as those classified as quadword interunit moves (e.g. a 64-bit move from an XMM register to the RBP register).
- It also so happens that:
  - any given build of GCC (of which GNAT is a part) detects, configures and fixes at build time to only emit one of those assembly syntaxes;
  - quadword interunit moves appear to be unavoidably emitted when building GNAT 12.x on x86_64-darwin; and
  - the only available GNAT bootstrap compilers for x86_64-darwin all emit quadword interunit moves with syntax recognised by the Clang integrated assembler that now comes with Xcode, but not the `cctools-port` assembler.
- All of this meant that, in order for Nixpkgs to build GNAT 12.x on and for x86_64-darwin, the build environment to build GNAT 12.x needed to have _both_ of the above assemblers in it:
  - the Clang integrated assembler so that, when building the GNAT stage 1 compiler from source, the assembly emitted by the GNAT bootstrap compiler can actually be assembled into object code; and
  - the `cctools-port` assembler so that the stage 1, 2 and 3 compilers can all be built to emit assembly for _that_ assembler (with stage 1 being built with the _other_ assembler, but stage 2 & 3 then being built with this one).
- The only way that I could see to get both assemblers and the GNAT bootstrap compiler into the build environment for GNAT 12.x was to:
  - change the stdenv so that its compiler was the GNAT bootstrap compiler package, and then
  - when on x86_64-darwin, have the GNAT bootstrap compiler package come with both of the above assemblers (via setting the `bintools` attribute at https://github.com/NixOS/nixpkgs/pull/209423/commits/f061f3ead59edc78259f537669afb15ac268e055#diff-ab5748dc9567516fefba8344056b51ec1866adeace380f46e58a7af3d619ea22R14500-R14502 to a custom package that actually contained both assemblers).